### ISSUE TITLE
[fsdp, model] Packed cu_seqlens detection misses TRL value-head critics (Qwen3.5)

### ISSUE THREAD

**Issue by liruiluo** [verl-project/verl#7931]
### System Info

- veRL source: [`3efe38c759c14622fd1b2c9e3679f2d02f86bdac`](https://github.com/verl-project/verl/commit/3efe38c759c14622fd1b2c9e3679f2d02f86bdac) (current `main` when checked).
- TRL source: `v0.27.0`, the version pinned by the FSDP extra at this veRL revision.
- Reproducer: Python 3.9.6, standard library only. It downloads and verifies public source files, then executes selected original functions/statements with recording stubs. It does **not** import the full frameworks, load a checkpoint, or run CUDA/training.

### Summary

When the value-model loader returns `trl.AutoModelForCausalLMWithValueHead`, the FSDP engine's packed-boundary capability check returns `False` even when the wrapped, patched Qwen3.5 model supports `cu_seqlens`.

With remove-padding enabled, the engine consequently omits both `cu_seqlens` and `cu_seqlens_cpu` from the critic's packed model inputs. TRL can transparently forward these arguments; the problem is that the engine does not supply them because it checks only the wrapper's explicit signature.

This is a missing-wrapper case in the boundary plumbing added by #6660, rather than a claim that its actor fix did not work.

### Relevant configuration and scope

- FSDP/FSDP2 `value_model` engine.
- Qwen3.5 underlying model, patched by veRL; `use_fused_kernels=False` for the path examined here.
- Critic implemented through the **TRL value-head wrapper**.
- `use_remove_padding=True`, with more than one independent sample packed into a microbatch.
- The capability check/input omission also applies at Ulysses SP size 1; SP > 1 is not required.

[`load_valuehead_model`](https://github.com/verl-project/verl/blob/3efe38c759c14622fd1b2c9e3679f2d02f86bdac/verl/utils/model.py#L641-L682) first attempts a native token-classification model before falling back to TRL. This report is specifically about the TRL branch; it does **not** assert that every current Qwen3.5 critic selects that branch or that all other value-head implementations/backends behave identically.

### Source trace

1. [`FSDPEngine._build_model_optimizer`](https://github.com/verl-project/verl/blob/3efe38c759c14622fd1b2c9e3679f2d02f86bdac/verl/workers/engine/fsdp/transformer_impl.py#L588-L596) probes only the returned outer module:

   ```python
   self.pass_packed_cu_seqlens = "cu_seqlens" in signature(module.forward).parameters
   ```

2. [TRL's value-head `forward`](https://github.com/huggingface/trl/blob/v0.27.0/trl/experimental/ppo/modeling_value_head.py#L703-L738) has `**kwargs`, but no explicit `cu_seqlens` parameter. It forwards those kwargs to `self.pretrained_model`:

   ```python
   def forward(self, input_ids=None, past_key_values=None,
               attention_mask=None, return_past_key_values=False, **kwargs):
       ...
       base_model_output = self.pretrained_model(
           input_ids=input_ids, attention_mask=attention_mask, **kwargs
       )
   ```

3. [`prepare_model_inputs`](https://github.com/verl-project/verl/blob/3efe38c759c14622fd1b2c9e3679f2d02f86bdac/verl/workers/engine/fsdp/transformer_impl.py#L1254-L1272) inserts the packed boundaries only if that capability flag is true. With the TRL wrapper in the text-only reproducer, the resulting keys are just `input_ids`, `attention_mask`, and `position_ids` (`attention_mask=None`).

4. The [patched Qwen3.5 decoder's linear-attention branch](https://github.com/verl-project/verl/blob/3efe38c759c14622fd1b2c9e3679f2d02f86bdac/verl/models/transformers/qwen3_5.py#L338-L359) consumes explicit `cu_seqlens`/`cu_seqlens_cpu`; it does not recover them from `position_ids` resets. Thus those isolation inputs do not reach that branch through the affected critic path.

[`FSDPEngineWithValueHead`](https://github.com/verl-project/verl/blob/3efe38c759c14622fd1b2c9e3679f2d02f86bdac/verl/workers/engine/fsdp/transformer_impl.py#L1594-L1599) inherits the LM engine's input preparation. V1 [`separate_async`](https://github.com/verl-project/verl/blob/3efe38c759c14622fd1b2c9e3679f2d02f86bdac/verl/trainer/ppo/v1/trainer_separate_async.py#L44-L84) also inherits the [base trainer's critic construction](https://github.com/verl-project/verl/blob/3efe38c759c14622fd1b2c9e3679f2d02f86bdac/verl/trainer/ppo/v1/trainer_base.py#L278-L338), so switching trainer scheduling modes does not replace this engine check. This is source tracing, not a completed V1 training reproduction.

### Reproduction

Save the following as `reproduce_wrapper_dispatch.py` and run:

```bash
python3 reproduce_wrapper_dispatch.py
```

The reproducer fetches three public source files, checks their SHA-256 hashes, and executes the **original** capability predicate, packed-input insertion statements, TRL forward method, and Qwen decoder method. Tensor/model operations are replaced with recording stubs to isolate argument routing. The actor function is used only for its signature. No task data or weights are needed.

<details>
<summary>Self-contained source-level reproducer</summary>

```python
"""Source-level dispatch regression; no model, torch, CUDA, or numerical claims."""

import ast
import copy
import hashlib
import inspect
from types import SimpleNamespace as NS
from urllib.request import urlopen

REV = "3efe38c759c14622fd1b2c9e3679f2d02f86bdac"
BASE = f"https://raw.githubusercontent.com/verl-project/verl/{REV}/"


def fetch(url, expected_sha256):
    with urlopen(url, timeout=45) as response:
        data = response.read()
    assert hashlib.sha256(data).hexdigest() == expected_sha256, url
    return ast.parse(data.decode())


engine = fetch(
    BASE + "verl/workers/engine/fsdp/transformer_impl.py",
    "79d8779fb2a0b8a1ac7b3ec85fba2521ac8cae079925a497c4366fd77f8abfbe",
)
qwen = fetch(
    BASE + "verl/models/transformers/qwen3_5.py",
    "a6abf55ac6dd846e2d8691d75eab09ef95885394ac84d532185f2ca78dacb99f",
)
trl = fetch(
    "https://raw.githubusercontent.com/huggingface/trl/v0.27.0/"
    "trl/experimental/ppo/modeling_value_head.py",
    "4c451b1139743e37a0d32da3046d64978627bab947567e149f574f1934e4e8ef",
)


def method(tree, name, owner=None):
    body = tree.body
    if owner:
        body = next(n for n in body if isinstance(n, ast.ClassDef) and n.name == owner).body
    return next(n for n in body if isinstance(n, ast.FunctionDef) and n.name == name)


def execute(nodes, namespace):
    nodes = ast.parse("from __future__ import annotations").body + copy.deepcopy(nodes)
    exec(compile(ast.fix_missing_locations(ast.Module(nodes, [])), "<source-extract>", "exec"), namespace)


def function(node):
    node = copy.deepcopy(node)
    node.decorator_list = []
    namespace = {"torch": NS(float32="float32")}
    execute([node], namespace)
    return namespace[node.name]


build = method(engine, "_build_model_optimizer", "FSDPEngine")
predicate = next(n for n in build.body if isinstance(n, ast.Try) and "pass_packed_cu_seqlens" in ast.unparse(n))
actor_forward = function(method(qwen, "forward_with_normal_backend"))
critic_forward = function(method(trl, "forward", "AutoModelForCausalLMWithValueHead"))


def capability(forward):
    state = NS()
    module = type("SignatureProbe", (), {"forward": forward})()
    execute([predicate], {"self": state, "module": module, "signature": inspect.signature})
    return state.pass_packed_cu_seqlens


actor_flag, critic_flag = capability(actor_forward), capability(critic_forward)
assert actor_flag is True and critic_flag is False
print("capability: underlying Qwen =", actor_flag, "; TRL wrapper =", critic_flag)

prepare = method(engine, "prepare_model_inputs", "FSDPEngineWithLMHead")
rmpad = next(n for n in prepare.body if isinstance(n, ast.If) and ast.unparse(n.test) == "use_remove_padding")
index = next(i for i, n in enumerate(rmpad.body) if isinstance(n, ast.Assign)
             and any(isinstance(t, ast.Name) and t.id == "model_inputs" for t in n.targets))
input_statements = rmpad.body[index:index + 2]
assert isinstance(input_statements[1], ast.If)
assert ast.unparse(input_statements[1].test) == "packed_cu_seqlens is not None and pass_packed_cu_seqlens"


class Offsets(list):
    def cpu(self):
        return self


offsets = Offsets([0, 4, 7])


def inputs(flag):
    namespace = dict(input_ids_rmpad=[[10, 11, 12, 13, 20, 21, 22]],
                     position_ids_rmpad=[[0, 1, 2, 3, 0, 1, 2]],
                     packed_cu_seqlens=offsets, pass_packed_cu_seqlens=flag,
                     output_args={"pad_size": 0})
    execute(input_statements, namespace)
    return namespace["model_inputs"]


critic_inputs = inputs(critic_flag)
assert "cu_seqlens" not in critic_inputs and "cu_seqlens_cpu" not in critic_inputs
print("critic model_inputs:", sorted(critic_inputs))


class TensorStub:
    device, dtype = "cpu", "float32"

    def squeeze(self, dim):
        return self


class HeadStub:
    summary = NS(weight=TensorStub())

    def __call__(self, hidden):
        return hidden


received = []


def recorder(**kwargs):
    received.append(kwargs)
    return NS(hidden_states=[TensorStub()], logits=TensorStub(), loss=None)


wrapper = NS(pretrained_model=recorder, v_head=HeadStub(), is_peft_model=False)
critic_forward(wrapper, **critic_inputs)
assert "cu_seqlens" not in received[-1]
print("current engine -> TRL -> underlying: boundary absent")
critic_forward(wrapper, **inputs(True))  # Control: explicitly supply the known-supported metadata.
assert received[-1]["cu_seqlens"] is offsets and received[-1]["cu_seqlens_cpu"] is offsets
print("explicit boundary -> same TRL forward -> underlying: preserved")

decoder = function(method(qwen, "qwen3_5_decoder_layer_forward"))
linear_calls = []
layer = NS(layer_type="linear_attention", input_layernorm=lambda x: x,
           post_attention_layernorm=lambda x: x, mlp=lambda x: 0,
           linear_attn=lambda **kwargs: linear_calls.append(kwargs) or 0)
decoder(layer, 0, (None, None), position_ids=[[0, 1, 2, 3, 0, 1, 2]])
assert linear_calls[-1]["cu_seqlens"] is None
print("decoder -> linear branch with position reset only: cu_seqlens = None")
print("PASS: dispatch regression reproduced; this is not a model/numerical test.")
```

</details>

Observed output (exit code 0):

```text
capability: underlying Qwen = True ; TRL wrapper = False
critic model_inputs: ['attention_mask', 'input_ids', 'position_ids']
current engine -> TRL -> underlying: boundary absent
explicit boundary -> same TRL forward -> underlying: preserved
decoder -> linear branch with position reset only: cu_seqlens = None
PASS: dispatch regression reproduced; this is not a model/numerical test.
```

The explicit-boundary control uses the same TRL forward body: when the arguments are supplied, both reach the underlying model unchanged. It does not constitute a numerical validation of a production fix.

### Expected behavior

A supported transparent value-head wrapper should not cause the engine to drop the sequence boundaries required by its underlying packed model. Independently packed samples need to remain isolated in the critic, including its linear-attention state.

A possible direction is to resolve capability through **known, verified transparent wrappers**, or expose an explicit packed-boundary capability contract. Accepting arbitrary `**kwargs` as proof of support would be unsafe for other models.

Suggested regression coverage:

- The patched model and its TRL value-head wrapper both receive identical packed boundary metadata.
- Models that do not support these arguments are not passed them indiscriminately.
- A real-model critic test compares the same sample alone versus packed after different preceding samples, in inference and training, with appropriate numerical tolerances.

### Related reports and validation limits

- #6660 is merged and adds the explicit-signature check as part of Qwen3.5 packing/SP support.
- #6780 reports the related Qwen3.5 packing/isolation problem in an actor/GRPO setup and was closed following a fix. That validation does not exercise this TRL critic wrapper.
- #7471 concerns delta-net kernel/block compatibility; this report concerns the outer value-head wrapper and engine capability detection.

The demonstrated defect here is **boundary-argument omission**. Real-checkpoint value differences, PPO-update differences, and downstream training-quality effects have not been measured by this reproducer and are not claimed in this report.

**Comment by wuxibin89** [comment]
@liruiluo Is this a trl bug specific to v0.27.0? Is it fixed in newer trl version? If so, then we should bump trl dependency.

**Comment by liruiluo** [comment]
@wuxibin89 Thanks for checking. This is not specific to TRL v0.27.0: the omission happens in veRL's capability detection, before TRL's forward is called.

I checked **v0.27.0, v0.28.0, and v1.12.0**. Their value-head `forward` implementations have the same AST: they accept `**kwargs`, have no explicit `cu_seqlens` parameter, and pass the kwargs to `self.pretrained_model`. See, for example, [v1.12.0's implementation](https://github.com/huggingface/trl/blob/59c4a8e104413fa9f4ca1a54eaf2ff93c0f299be/trl/experimental/ppo/modeling_value_head.py#L716-L751).

Re-running the issue's **source-level routing check** against those three implementations gives the same result:

| TRL version | Existing veRL probe on the wrapper | Explicitly supplied boundary tensors forwarded unchanged |
| --- | --- | --- |
| 0.27.0 | `False` | Yes |
| 0.28.0 | `False` | Yes |
| 1.12.0 | `False` | Yes |

This uses original source statements with recording doubles, not a new real-model or training test.

There is also an important latest-version caveat: **TRL v1.13.0 removes the PPO/value-head module**, via huggingface/trl#7020. It is therefore not a drop-in dependency bump for veRL's existing TRL value-head fallback; that path would need adaptation rather than gaining a fixed wrapper.

So upgrading TRL alone does not resolve this detection gap. #7932 addresses it in veRL by inspecting the underlying explicit capability only for the known transparent value-head wrapper, while continuing to call that wrapper and preserving the guard against arbitrary `**kwargs` models.
